radixdlt-core/radixdlt/src/main/java/com/radixdlt/client/store/berkeley/BerkeleyClientApiStore.java

```diff
@@ -0,0 +1,296 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied.  See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+package com.radixdlt.client.store.berkeley;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import com.google.inject.Inject;
+import com.radixdlt.atom.SpunParticle;
+import com.radixdlt.atommodel.tokens.StakedTokensParticle;
+import com.radixdlt.atommodel.tokens.TransferrableTokensParticle;
+import com.radixdlt.atommodel.tokens.UnallocatedTokensParticle;
+import com.radixdlt.client.store.ClientApiStore;
+import com.radixdlt.client.store.ClientApiStoreException;
+import com.radixdlt.client.store.TokenBalance;
+import com.radixdlt.constraintmachine.Particle;
+import com.radixdlt.constraintmachine.Spin;
+import com.radixdlt.identifiers.RadixAddress;
+import com.radixdlt.serialization.DeserializeException;
+import com.radixdlt.serialization.Serialization;
+import com.radixdlt.store.DatabaseEnvironment;
+import com.radixdlt.store.LedgerEntryStore;
+import com.radixdlt.utils.RadixConstants;
+import com.radixdlt.utils.functional.Result;
+import com.sleepycat.je.Cursor;
+import com.sleepycat.je.Database;
+import com.sleepycat.je.DatabaseConfig;
+import com.sleepycat.je.DatabaseEntry;
+import com.sleepycat.je.OperationStatus;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ExecutorService;
+
+import static com.google.common.primitives.UnsignedBytes.lexicographicalComparator;
+import static com.radixdlt.serialization.DsonOutput.Output;
+import static com.radixdlt.utils.ThreadFactories.daemonThreads;
+
+import static java.util.concurrent.Executors.newSingleThreadExecutor;
+
+public class BerkeleyClientApiStore implements ClientApiStore {
+	private static final Logger log = LogManager.getLogger();
+
+	private static final String EXECUTED_TRANSACTIONS_DB = "radix.executed_transactions_db";
+	private static final String BALANCE_DB = "radix.balance_db";
+
+	private final DatabaseEnvironment dbEnv;
+	private final LedgerEntryStore store;
+	private final Serialization serialization;
+	private final StackingCollector<SpunParticle> particleCollector = StackingCollector.create();
+	private final ExecutorService executorService = newSingleThreadExecutor(daemonThreads("ClientApiStore"));
+
+	private Database executedTransactionsDatabase;
+	private Database balanceDatabase;
+
+	@Inject
+	public BerkeleyClientApiStore(DatabaseEnvironment dbEnv, LedgerEntryStore store, Serialization serialization) {
+		this.dbEnv = dbEnv;
+		this.store = store;
+		this.serialization = serialization;
+
+		open();
+	}
+
+	@Override
+	public Result<List<TokenBalance>> getTokenBalances(RadixAddress address) {
```

Comment:
This should only return one entry per token (RRI), with staked amount subtracted, only immediately spendable tokens :).

Reply:
Fixed